sql/core/src/main/java/org/apache/spark/sql/sources/v2/SupportsBatchRead.java

```diff
@@ -20,14 +20,27 @@
 import org.apache.spark.annotation.Evolving;
 import org.apache.spark.sql.sources.v2.reader.Scan;
 import org.apache.spark.sql.sources.v2.reader.ScanBuilder;
+import org.apache.spark.sql.types.StructType;
 
 /**
- * An empty mix-in interface for {@link Table}, to indicate this table supports batch scan.
- * <p>
- * If a {@link Table} implements this interface, its {@link Table#newScanBuilder(DataSourceOptions)}
- * must return a {@link ScanBuilder} that builds {@link Scan} with {@link Scan#toBatch()}
- * implemented.
- * </p>
+ * A mix-in interface for {@link Table} to provide data reading ability of batch processing.
  */
 @Evolving
-public interface SupportsBatchRead extends Table { }
+public interface SupportsBatchRead extends Table {
+
+  /**
+   * Returns the schema of this table.
+   */
+  StructType schema();
```

Comment:
I'm not sure about this. Maybe it's ok to leave `schema` in `Table`, and asks write-only table to report schema as empty.

Reply:
To me, +1 for the current change.

Reply:
I think that `schema` should be a method on `Table`. Write-only tables still need to access the table's schema to validate a write.

Reply:
@rdblue . Validating is one of the important use cases, but there are another use cases. We've already received the request previously. 
> The schema is defined by the dataframe itself, not by the data source, i.e. it should be extracted from df.schema and not by source.createReader 

- http://apache-spark-developers-list.1001551.n3.nabble.com/Possible-bug-in-DatasourceV2-td25343.html

Reply:
@dongjoon-hyun, the problem that you pointed out was that the schema shouldn't require a reader. 
That's what I'm saying here, too: the table should have a schema and it shouldn't need to implement the read path to validate a write using the table schema.

Reply:
I think `Table` should also have a `schema`. I agree that it is named, but it is also a data source or sink of some kind, with a schema. I don't see the value in moving the schema to a different interface, and I think that moving the schema to an interface specific to the read path is worse because it causes the problem that a table must be readable to be writable.

Reply:
@rdblue what about the use case in http://apache-spark-developers-list.1001551.n3.nabble.com/Possible-bug-in-DatasourceV2-td25343.html ?

i.e. the data source can write data of any schema, and it's not readable, it has no schema?

Do you mean we should ask data source to return an empty schema for this case?

Reply:
or we treat it as an invalid data source, and ask users to use `df.foreach` instead?

Reply:
@cloud-fan, in that case, we're talking about a change to how we apply validation rules. I've proposed a way to customize the validation rules for a data source, using capabilities.

For example, we could add a capability that instructs Spark to not validate the schema for a write. I don't think that's very useful, so I think it is more likely that we would add a capability that instructs Spark to loosen the validation in some way -- like the one I've already suggested that allows omitting columns that are optional.

Reply:
I should also note that writing to a source that can't be read and doesn't have a consistent schema is far outside the primary uses of this API. Keep in mind that we should design primarily for tables with schemas; it's great to make that use case possible, but we should not alter the design too much to do it.

Reply:
I'd moving the schema stuff to a new single interface. That would be better for Single Responsibility Principle. @rdblue , in that case, we are good, right? Is there any other concerns?
> I don't see the value in moving the schema to a different interface, and I think that moving the schema to an interface specific to the read path is worse because it causes the problem that a table must be readable to be writable.

Reply:
I think that the use case for a table with no schema is extremely narrow, if not laughably unlikely.

Silly use cases should not cause us to make changes to an API when there is a reasonable alternative: a source with no schema should return an empty schema and signal that it wants to disable write validation (using a capability).

There are two advantages to this approach:

1. Implementations need to implement `schema` and validation rules are enabled by default. If an implementation needs to remember to add a `HasSchema` interface to turn on validation rules, then the default is wrong. The result would be that people less familiar with the API will not have validated writes, and that's a problem.
2. The API is simpler and easier to understand.

I'm -1 on moving this from `Table`.

Reply:
Okay. You think so. But, obviously, we already wasted a lot of time for this `laughably unlikely` stuff, didn't we?

Reply:
Did we? What are the examples of sources that are write-only and don't need schema validation? I could be wrong here, but I didn't think that there were many.